### PR TITLE
dom: Add types to document-has-selection

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Add types to `document-has-selection` and its three dependencies.

Some runtime changes are necessary to appease TypeScript. I've annotated each with the reason it is needed and why I think it's safe.

Part of #18838 

## How has this been tested?
Unit tests and e2e tests.

## Types of changes
Non-breaking changes.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/dom/src/dom/document-has-selection.js
Comment (sarayourfriend) [inline packages/dom/src/dom/document-has-selection.js:18]: If a document has no `activeElement` then it cannot have a selection. Furthermore, `isTextField` and `isNumberInput` access the passed in element completely unguarded, so the assumption here was always that `activeElement` was defined.

File: packages/dom/src/dom/document-has-text-selection.js
Comment (sarayourfriend) [inline packages/dom/src/dom/document-has-text-selection.js:22]: `range` is not a boolean so we cast it to a boolean to appease the return value.

File: packages/dom/src/dom/is-text-field.js
Comment (sarayourfriend) [inline packages/dom/src/dom/is-text-field.js]: `nonTextInputs` is a `string[]` and because `element.type` could be `undefined | string` it doesn't fit in the box of a `string[]`. Checking for it's presence first turns it into a `string` which we can then pass to the `includes` function on the next line.

File: packages/dom/README.md
Comment (ellatrix) [inline packages/dom/README.md]: Pretty new to this. Why is `Partial` needed?

Comment (sarayourfriend) [reply packages/dom/README.md]: `Partial` is needed because the `Element` itself could be any HTML element that we're testing to check to see if it's an input, textarea or `contentEditable` Element. I wasn't sure how else to express that the `Input` part is optional and I think this is what @sirreal suggested as a potential solution to this problem (though I may be mis remembering, I can't find the original comment where that was suggested.

Comment (sirreal) [reply packages/dom/README.md]: I believe this was introduced because we have some type that `HTMLInputElement` extends, but we don't know that it's an `HTMLInputElement`.
`Partial<HTMLInputElement>` doesn't seem ideal to me, it's typically used when we have an interface with optional properties. This usage is really that we have an `Element` (I believe, could be another type) and want to use it safely if it satisfies `HTMLInputElement`.

In my opinion, a better approach would be to accept the element type we expect to pass in (`Element`?), and use a type guard to narrow our element to our desired type. I implemented a type guard like this in #29220 to handle this type of case:

https://github.com/WordPress/gutenberg/blob/29601bd2a0427037dde074a376d4b23e2ae61460/packages/dom/src/dom.js#L83-L89

@sarayourfriend I believe you're thinking of this comment: https://github.com/WordPress/gutenberg/pull/30030#discussion_r598714524

Comment (ellatrix) [reply packages/dom/README.md]: Right, maybe just `HTMLElement`? We should allow any element to be checked with this function.

Comment (sarayourfriend) [reply packages/dom/README.md]: Aha, yes, I did try to create a type guard but got stuck on `contentEditable`... I'll revisit this though and see if I can make it work with a type guard.

Comment (ellatrix) [reply packages/dom/README.md]: Meaning: sometimes the flexibility is nice so you don't have to check the type before calling a function that's being strict about the type. I think this is generally good for all functions in this package.

Comment (sarayourfriend) [reply packages/dom/README.md]: Looks like I just needed a couple type casts and it works out well even without the type guard.

File: packages/dom/README.md
Comment (ellatrix) [inline packages/dom/README.md]: `HTMLElement` doesn't work?

File: packages/dom/src/dom/is-text-field.js
Comment (ellatrix) [inline packages/dom/src/dom/is-text-field.js:6]: What's this for?

File: packages/dom/src/dom/is-text-field.js
Comment (ellatrix) [inline packages/dom/src/dom/is-text-field.js]: Does it not detect the `element.nodeName === 'INPUT'` check? What if you check `typeof element === HTMLInputElement`?

File: packages/dom/README.md
Comment (sarayourfriend) [reply packages/dom/README.md]: It seems like `HTMLElement` is a narrow type than is necessary. From the documentation on `Element`:

> `Element` is the most general base class from which all objects in a `Document` inherit. It only has methods and properties common to all kinds of elements. More specific classes inherit from `Element`

Whereas `HTMLElement` is:

> Any HTML element. Some elements directly implement this interface, while others implement it via an interface that inherits it.

Maybe it should be left as `HTMLElement` instead? 🤔 I'm not sure 😬 I'll defer to @sirreal again on this one, sorry to pull you back in Jon!

File: packages/dom/src/dom/is-text-field.js
Comment (sarayourfriend) [reply packages/dom/src/dom/is-text-field.js:6]: Whoops, this was left over from when I tried to make this a type guard, but it can be removed now. For the record, the type guard syntax sets off `jsdoc/valid-types`:

```
/**
 * @param {Element} element
 * @return {element is HTMLInputElement | HTMLTextAreaElement}
 */
```

We frequently have to disable `jsdoc/valid-types` to be able to use valid TypeScript syntax in the curly braces.

File: packages/dom/src/dom/is-text-field.js
Comment (sarayourfriend) [reply packages/dom/src/dom/is-text-field.js]: It doesn't, we'd need a type guard for it to work that way. I'll see if `typeof` works though! Alternatively I could introduce a function `isHTMLInputElement`:

```js
/**
 * @param {Element | null | undefined} element
 * @return {element is HTMLInputElement}
 */
function isHTMLInputElement( element ) {
  return element && element.nodeName === 'INPUT';
}
```

That would definitely work.

Comment (sarayourfriend) [reply packages/dom/src/dom/is-text-field.js]: I get this error when I try to use `typeof`:

`This condition will always return 'false' since the types 'string' and '{ new (): HTMLInputElement; prototype: HTMLInputElement; }' have no overlap.`

`eslint` also complains that `HTMLInputElement` is undefined 😞 I'll add a type guard?

File: packages/dom/README.md
Comment (ellatrix) [reply packages/dom/README.md]: AS far as I know, `Element` can also be `SVGElement` and perhaps custom types. Just `Element` sounds good though. 🤷‍♀️

File: packages/dom/src/dom/is-text-field.js
Comment (ellatrix) [reply packages/dom/src/dom/is-text-field.js]: Up to you. :) I find these inline comments a bit annoying to be honest, but at this point I'm nit picking.

Comment (sarayourfriend) [reply packages/dom/src/dom/is-text-field.js]: Well I could just convert it to TypeScript and the inline comments would go away 😝 

I'll do the type guard though, we should have something like that anyhow I think, it'll be useful in other cases as well I'm sure.

File: packages/dom/src/dom/is-html-input-element.js
Comment (ellatrix) [inline packages/dom/src/dom/is-html-input-element.js:1]: Why is it disabled here and enabled below?

File: packages/dom/src/dom/is-html-input-element.js
Comment (ellatrix) [inline packages/dom/src/dom/is-html-input-element.js]: Are we planning to pass `null` or `undefined`? I see in `isNumberInput` we already require an `Element`, so can't we require it here?

File: packages/dom/src/dom/is-html-input-element.js
Comment (sarayourfriend) [reply packages/dom/src/dom/is-html-input-element.js:1]: Because the `element is HTMLInputElement` syntax isn't considered a "valid type" by the JSDoc parser (doctrine) that the eslint JSDoc rules use. If you search the codebase you'll see that we've unfortunately had to disable these all over the place.

File: packages/dom/src/dom/is-html-input-element.js
Comment (sarayourfriend) [reply packages/dom/src/dom/is-html-input-element.js]: If this function is used in other places it might be helpful to also handle `null` and `undefined` correctly. On the other hand, might be a YAGNI situation. I'll leave it up to you.

**Review by ellatrix — APPROVED** [review]
I'm not super comfortable with the typing and the bigger picture/effort, so maybe good to have a second blessing. 😄

**Comment by sarayourfriend** [comment]
I requested review from @tyxla who's been helping with these recently :)

**Review by tyxla — APPROVED** [review]
Looks great to me! 👍 

I wonder why the lint check failed, though - I didn't see any related failures.

**Comment by sarayourfriend** [comment]
Not sure @tyxla. I rebased, hopefully that will resolve the error coming from the interface package 🤞

**Comment by sarayourfriend** [comment]
Well after rebasing now the e2e tests are failing 😞 I'm digging into it now.

I think the test is failing on `trunk` as I'm unable to get the test to pass locally on `trunk` or my branch.

File: packages/dom/README.md
Comment (sirreal) [reply packages/dom/README.md]: I'd say it's good practice to generalize parameter types as much as possible. That will limit what we can do with the values internally. Externally, it will allow these functions to be used in more places.

For example, this function checks the [`nodeName` property](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeName), which exists on [`Node`](https://developer.mozilla.org/en-US/docs/Web/API/Node) ([from which Element inherits](https://developer.mozilla.org/en-US/docs/Web/API/Element)):

```ts
export default function isHTMLInputElement( node: Node ): node is HTMLInputElement {
	return node.nodeName === 'INPUT';
}
```

That seems ideal to me. The parameter is exactly as strict as it must be for it to be used internally.

File: packages/dom/src/dom/is-html-input-element.js
Comment (sirreal) [reply packages/dom/src/dom/is-html-input-element.js]: It's easy to widen types at any point, but a breaking change to narrow them. If this is ever released accepting `X | null | undefined` , that's more difficult to narrow to `X` than the other way around. A function that accepts `X` can start the wider accepting `X | null | undefined` at any point.

I agree, it might be handy to accept `null` for taking input from things like [`Document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) or [`Node.parentNode`](https://developer.mozilla.org/en-US/docs/Web/API/Node/parentNode) that may return `null`. However, I'd avoid widening the type until it's absolutely necessary.

Comment (sirreal) [inline packages/dom/src/dom/is-html-input-element.js]: I [mentioned above](https://github.com/WordPress/gutenberg/pull/30386/files#r607019420), but it seems like this could accept values of `Node` type:
```suggestion
 * @param {Node | null | undefined} element
```

File: packages/dom/src/dom/is-number-input.js
Comment (sirreal) [inline packages/dom/src/dom/is-number-input.js]: Our `isHTMLInputElement` function seems perfect for this:
```suggestion
	return isHTMLInputElement( element ) && !! element.valueAsNumber;
```

File: packages/dom/src/dom/is-number-input.js
Comment (sirreal) [inline packages/dom/src/dom/is-number-input.js]: This should probably be whatever `isHTMLInputElement` accepts.
```suggestion
 * @param {Node} element The HTML element.
```

File: packages/dom/src/dom/is-number-input.js
Comment (sirreal) [inline packages/dom/src/dom/is-number-input.js]: We've got the same type guard as `isHTMLInputElement` here, let's take advantage:
```suggestion
 * @return {element is HTMLInputElement} True if the element is input and holds a number.
```

File: packages/dom/src/dom/is-text-field.js
Comment (sirreal) [inline packages/dom/src/dom/is-text-field.js]: Slight preference for moving the assertion from the destructuring assignment above to here. We should try to keep assertions as small and localized as possible:

```suggestion
		/** @type { HTMLElement } */ ( element ).contentEditable === 'true'
```

----

Interesting aside, I just learned that `contentEditable` can also be `"inherit"`. Checking [the `boolean` `isContentEditable` property](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/isContentEditable) is recommended for checking the computed value (in case of `inherit`).

Out of scope for this but worth mentioning 🙂

File: packages/dom/src/dom/is-text-field.js
Comment (sirreal) [inline packages/dom/src/dom/is-text-field.js]: Let's keep the assertion as local as possible:
```suggestion
	const { nodeName } = element;
```

File: packages/dom/src/dom/is-text-field.js
Comment (sirreal) [inline packages/dom/src/dom/is-text-field.js]: I think we could accept `Node` here again.

File: packages/dom/src/dom/is-text-field.js
Comment (sirreal) [inline packages/dom/src/dom/is-text-field.js]: This boolean result narrows us to at least `HTMLElement`, right?
```suggestion
 * @return {element is HTMLElement} True if the element is an text field, false if not.
```

**Review by sirreal — COMMENTED** [review]
Good stuff here, thanks! There are some interesting API design questions around accepting null | undefined values.

I've left some notes but nothing blocking.

**Pull request merged**
